### PR TITLE
Send confirmation email

### PR DESCRIPTION
**Pull request by whit537**
Part of #4305.

### Requirements

(7) and (8) from the [workflow](https://github.com/gratipay/gratipay.com/issues/4297#issuecomment-273900610):

> (7) receive email
> (8) click link

### Todo

- [x] fix these three failing tests
- [x] write Python API w/ tests & docs—should put mail on the queue
  - [x] ~~make sure we don't allow connecting the same package to two accounts~~<br>that happens on `finish-claim`, not `start-claim`
  - [x] make sure we're validating the address, can't send to just anyone
  - [x] extend `start_email_verification` to take packages
    - [x] what about packages in the except case?
  - [x] extend `add_email` to take packages
- [x] what happens to old `claims` when nonce is updated out from under it?
- [x] modify `modify.json.spt` w/ tests—should call into Python
  - [x] consider `LazyResponse`
- [x] write the email template
- [x] ~~should notify others listed as maintainers?~~ no, too much for now
- [x] can we inject via npm package name?
- [x] don't say we're connecting when we're not (ancillary accounts)
- [x] ~~should be able to start a verification for a different package w/ same email and have it not clobber the first (added? in addition?)~~ https://github.com/gratipay/gratipay.com/pull/4335#issuecomment-290554628

### CODE REVIEW

**Comment by whit537** [comment]
Rebased, was f3e657ffef9b9f07855195467189a366ccbe69e2.

**Comment by whit537** [comment]
Gosh, this would be easier to implement if we kept it to two steps:

- add email
- claim package

But that makes it harder for users.

**Comment by whit537** [comment]
The natural thing for a user is to claim the package and add the email in one fell swoop.

**Comment by whit537** [comment]
Our email infrastructure is pretty squishy. I started really getting into it yesterday, and I'm trying to decide how deep to go.

**Comment by whit537** [comment]
Rebased.

**Comment by whit537** [comment]
> I started really getting into it yesterday, and I'm trying to decide how deep to go.

Went pretty deep! 😁

**Comment by whit537** [comment]
Alright, what the heck am I doing here?

**Comment by whit537** [comment]
What's the end result here? What behavior?

/me reviews ticket and parent ...

**Comment by whit537** [comment]
screen shot 2017-03-22 at 1 49 47 pm

**Comment by whit537** [comment]
(I updated the ticket description with requirements.)

**Comment by whit537** [comment]
Blorg. Right. Gotta think through adding email vs. mailing before adding. Current infrastructure assumes ... wait, does it? Assume we're only ever sending to `participant.email_address`? But then how does adding non-primary emails work?

**Comment by whit537** [comment]
Ah ... is the `To` specific in the `context` in that case?

**Comment by whit537** [comment]
[Yeah](https://github.com/gratipay/gratipay.com/blob/2065/gratipay/models/participant/email.py#L108), [that's](https://github.com/gratipay/gratipay.com/blob/2065/gratipay/email.py#L152) [it](https://github.com/gratipay/gratipay.com/blob/2065/gratipay/email.py#L170).

**Comment by whit537** [comment]
Existing [`verification_notice`](https://github.com/gratipay/gratipay.com/blob/2065/emails/verification_notice.spt) should be good enough for the package claiming case as well.

**Comment by whit537** [comment]
We can mostly use the existing `add_email` flow. The caveat is that we should send a mail even when the email already exists.

**Comment by whit537** [comment]
I think even the existing `verification` template will work, we can just say "connect **npm/aspen and chad@zetaweb.com** to ~whit537" as appropriate.

**Comment by whit537** [comment]
Gotta figure out how nonces are gonna work and how they will relate to packages. I think we may want a new table to track how nonces relate to packages, because nonces can either represent zero or one package, and down the road we're thinking they can also represent N packages in the case where someone wants to link multiple packages at once.

**Comment by whit537** [comment]
But first let's get these test fixed ...

**Comment by whit537** [comment]
Harumph. The trick here is to allow for sending a verification for package ownership even when the address itself is already verified, *and* not trip on our try/except/recurse log in `start_email_verification`.

**Comment by whit537** [comment]
Topped up. Looking at #4359 &rarr; #4322 before proceeding further ...

**Comment by whit537** [comment]
Returning from #4391 ...

**Comment by whit537** [comment]
Rebased to pick up #4391.

**Comment by whit537** [comment]
Alright, so I'm pretty sure the try/except/recurse wants to be tighter in, only around nonce creation vs. package linking. What else?

**Comment by whit537** [comment]
Alright, so what are the conditions here? I'm to the point where I need to validate `email, *packages` when there's always an email and there may or may not be `packages`.

**Comment by whit537** [comment]
We also need to think about package claiming across accounts.

|  | not claimed | half-claimed | already claimed |
|-------|--------|----------|----------|
| email |   |    |   |
| package (each) |   |   |   |

**Comment by whit537** [comment]
If there aren't packages, we want to keep doing what we've been doing: block if the email is already verified either by the participant in question or someone else, or if the participant already has 10 emails.

If there _are_ packages, ... then what?

**Comment by whit537** [comment]
Theeeennnnnn .......

**Comment by whit537** [comment]
Then we want to block if:

- the email address is already verified for someone else
- the email address is not on file for all of the packages
- ...? is that it?

**Comment by whit537** [comment]
I've confirmed that 2037a2f2cef199e493cb50818db80be30e6224ba properly extracts i18n messages, by running `make i18n` and inspecting `i18n/core.pot`.

**Comment by whit537** [comment]
Okay! I've got this pretty close. I'm going to start working on the email copy. 😀

**Comment by whit537** [comment]
Okay, phew! Email template in de1a0a32399173bc027b9242ab5da67197856a11. Let's see what it looks like live! 😮

After that there's one more edge case to consider:

> should be able to start a verification for a different package w/ same email and have it not clobber the first (added? in addition?)

**Comment by whit537** [comment]
💃

screen shot 2017-03-30 at 5 19 38 pm

**Comment by whit537** [comment]
screen shot 2017-03-30 at 5 21 46 pm

**Comment by whit537** [comment]
However, I'm not getting the actual verification link on chad+foo@zetaweb.com. Hmm ... :thinking:

**Comment by whit537** [comment]
Oh. Probably because I entered `chad+foo@example.com`. 😊

**Comment by whit537** [comment]
💃 

screen shot 2017-03-30 at 5 29 34 pm

**Comment by whit537** [comment]
screen shot 2017-03-30 at 5 30 11 pm

**Comment by whit537** [comment]
> After that there's one more edge case to consider:
>
>> should be able to start a verification for a different package w/ same email and have it not clobber the first (added? in addition?)

Actually, I think clobbering is fine to start with here. We've got all the wiring we need to implement a UI for verifying multiple packages at once, which will reduce the incentive to trigger multiple separate verification flows in parallel. I'm comfortable erroring out the first flows in that case. Do we do that?

**Comment by whit537** [comment]
Alright, started a test but it's bleeding into the next task, which is modifying the `verify.html` endpoint to receive package claims in addition to address verifications, so let's save that for the next PR.

```diff
diff --git a/tests/py/test_email.py b/tests/py/test_email.py
index 07ce431..292d764 100644
--- a/tests/py/test_email.py
+++ b/tests/py/test_email.py
@@ -286,6 +286,22 @@ class TestEndpoints(Alice):
         assert response.code == 400
         assert self.db.all('select package_id from claims order by package_id') == []
 
+    def test_starting_a_second_claim_clobbers_the_first(self):
+        foo = self.make_package()
+        start = lambda: self.hit_email_spt( 'start-verification'
+                                          , 'alice@example.com'
+                                          , package_ids=[foo.id]
+                                           )
+        start()
+        old_nonce = self.alice.get_email('alice@example.com').nonce
+        start()
+        new_nonce = self.alice.get_email('alice@example.com').nonce
+        one = self.verify_email('alice@example.com', old_nonce).body
+        two = self.verify_email('alice@example.com', new_nonce).body
+        import pdb; pdb.set_trace()
+        assert 'Bad Info' in one
+        assert 'Success!' in two
+
 
 class TestFunctions(Alice):
 
```

**Comment by whit537** [comment]
And with that, I am submitting this for review!

Ping @rohitpaulk @nobodxbodon @mattbk @JessaWitzel @aandis @clone1018.

This is the next PR in our long march towards claiming npm packages (#4305). How does it look? :-)

**Comment by whit537** [comment]
Rebased, was de1a0a32399173bc027b9242ab5da67197856a11.

**Comment by kaguillera** [comment]
LGTM...off to the second reviewr

**Comment by whit537** [comment]
screen shot 2017-04-12 at 5 32 25 pm

File: gratipay/models/participant/email.py [hunk 1]
```diff
@@ -41,94 +41,172 @@ class Email(object):
...
-                                     )
+        return nonce
+
+
+    def start_package_claims(self, c, nonce, *packages):
+        """Takes a cursor, nonce and list of packages, inserts into ``claims``
+        and returns ``None``.
+        """
+        if not packages:
+            return
```
Comment (dowski) [inline gratipay/models/participant/email.py]: Is this a normal condition? If not, consider logging or throwing an exception. Seems odd to start a package claim on an empty list of packages.

File: gratipay/models/participant/email.py [hunk 2]
```diff
@@ -41,94 +41,172 @@ class Email(object):
...
+        VALUES, values = [], []
```
Comment (dowski) [inline gratipay/models/participant/email.py]: Is this Gratipay convention, to use CAPS for format strings? Normally I'd expect that to be a constant.

File: gratipay/models/participant/email.py [hunk 3]
```diff
@@ -41,94 +41,172 @@ class Email(object):
...
+        for p in packages:
+            VALUES.append('(%s, %s)')
+            values += [nonce, p.id]
+        VALUES = ', '.join(VALUES)
```
Comment (dowski) [inline gratipay/models/participant/email.py]: It seems tricky to switch types of a variable from line to line. Maybe that's the Java coder in me though. ;-)

File: tests/py/test_email.py [hunk 4]
```diff
@@ -20,14 +22,35 @@ def setUp(self):
...
 class TestEndpoints(Alice):
 
-    def hit_email_spt(self, action, address, user='alice', should_fail=False):
+    def hit_email_spt(self, action, address, user='alice', package_ids=[], should_fail=False):
         f = self.client.PxST if should_fail else self.client.POST
-        data = {'action': action, 'address': address}
-        headers = {b'HTTP_ACCEPT_LANGUAGE': b'en'}
-        return f('/~alice/emails/modify.json', data, auth_as=user, **headers)
+
+        # Hack to work around Aspen test client limitations.
```
Comment (dowski) [inline tests/py/test_email.py]: What limitation?

File: emails/verification.spt [hunk 5]
```diff
@@ -1,16 +1,63 @@
 {{ _("Connect to {0} on Gratipay?", username) }}
 
 [---] text/html
-{{ _("We've received a request to connect {0} to the {1} account on Gratipay. Sound familiar?",
-     ('<b>%s</b>'|safe) % email,
-     ('<b><a href="https://gratipay.com/~{0}/">{0}</a></b>'|safe).format(username)) }}
+{% if new_email_verified %}
+{{ ngettext( "We've received a request to connect the {package_name} npm package to the "
+             "{username} account on Gratipay. Sound familiar?"
```
Comment (dowski) [inline emails/verification.spt]: What if it doesn't sound familiar?

What if it does, for that matter? I don't really know what I would do if I received an email like this.

Am I supposed to click the package name?

File: gratipay/models/participant/email.py [hunk 6]
```diff
@@ -41,94 +41,172 @@ class Email(object):
...
+        c.run('INSERT INTO claims (nonce, package_id) VALUES ' + VALUES, values)
```
Comment (dowski) [inline gratipay/models/participant/email.py]: I'm generally unfamiliar with the Gratipay DB environment. Does anything exist to clean up claims that lay around for a long time?

File: emails/verification.spt (hunk 5 above)
Comment (whit537) [inline emails/verification.spt]: Down on line 32 there's a "Yes, proceed!" button. Here's what it looks like:

screen shot 2017-04-13 at 8 41 55 am

Comment (dowski) [inline emails/verification.spt]: Ah, woops - didn't see that! Thanks!

File: tests/py/test_email.py (hunk 4 above)
Comment (whit537) [inline tests/py/test_email.py]: Comment verbosified in 136591a2074cb902df7853de632b87cbff6986ef.

File: gratipay/models/participant/email.py (hunk 3 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: Addressed in 2c2a3a51116cf617e472431178f3cc436822a8bc. For large inputs the `', '.join` idiom is more performant than string concatenation due to Python implementation details, but our inputs here are not large (order of magnitude 100 packages max to be claimed at once).

File: gratipay/models/participant/email.py (hunk 1 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: Done in 2c2a3a51116cf617e472431178f3cc436822a8bc. 👍

File: gratipay/models/participant/email.py (hunk 3 above)
Comment (dowski) [inline gratipay/models/participant/email.py]: FWIW, I didn't have a problem with .join(), it was just my internal type checker twitching at VALUES starting as a list of strings and then getting reassigned to a string.

File: gratipay/models/participant/email.py [hunk 7]
```diff
@@ -177,16 +178,20 @@ def get_email_verification_nonce(self, c, email):
...
+            raise NoPackages()
+
+        # We want to make a single db call to insert all claims, so we need to
+        # do a little SQL construction. Do it in such a way that we still avoid
+        # Python string interpolation (~= SQLi vector).
+
+        extra_sql, values = '', []
         for p in packages:
-            VALUES.append('(%s, %s)')
+            extra_sql += ' (%s, %s)'
```
Comment (dowski) [inline gratipay/models/participant/email.py]: Looks like maybe this isn't going to be delimited by commas anymore?

" (%s, %s) (%s, %s) (%s, %s)"

Maybe that's OK for Postgres though?

Comment (whit537) [inline gratipay/models/participant/email.py]: Good point. The few tests I ran passed. Lessee here ...

Comment (whit537) [inline gratipay/models/participant/email.py]: Yay CI! :-)

```
    def execute(self, query, vars=None):
        self.Record = None
>       return super(NamedTupleCursor, self).execute(query, vars)
E       ProgrammingError: syntax error at or near "("
E       LINE 1: ...LUES ('208503b2-9283-4bc8-ab51-6ed60c28c9a3', 30) ('208503b2...
E  
```

https://travis-ci.org/gratipay/gratipay.com/builds/221772022#L2879-L2884

Fixing ...

Comment (whit537) [inline gratipay/models/participant/email.py]: You okay with 992698534d07424fd60f123bf1c7a040d54f5471?

File: gratipay/models/participant/email.py (hunk 6 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: No! Reticketed as https://github.com/gratipay/gratipay.com/issues/4408.

File: gratipay/models/participant/email.py (hunk 3 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: P.S. @kaguillera thinks this _is_ the Java coder in you. 😛

**Comment by whit537** [comment]
With https://github.com/gratipay/inside.gratipay.com/issues/1038#issuecomment-293990454 you should be cleared to merge, @dowski. Got a green button?

**Comment by whit537** [comment]
Woo-hoo! 💃

**Pull request merged**